### PR TITLE
audio/out: add AudioUnit output driver for iOS

### PR DESCRIPTION
**Pull request by tmm1**
iOS only has a small subset of the CoreAudio APIs available on macOS.

This PR adds a new audio output driver for iOS, which re-uses some of the CoreAudio helpers along with the AudioUnit and AVAudioSession APIs that are available only on iOS.

### CODE REVIEW

File: audio/out/ao_audiounit.m [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+    }
+
+    err = AudioComponentInstanceNew(comp, &(p->audio_unit));
+    CHECK_CA_ERROR("unable to open audio component");
+
+    err = AudioUnitInitialize(p->audio_unit);
+    CHECK_CA_ERROR_L(coreaudio_error_component,
+                     "unable to initialize audio unit");
+
+    if (!af_fmt_is_spdif(ao->format) && instance.outputNumberOfChannels > 2) {
```
Comment (wm4) [inline audio/out/ao_audiounit.m]: Maybe invert the condition and switch the if/else branches.

File: audio/out/ao_audiounit.m [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+        if (instance.currentRoute.outputs.firstObject.channels.count == 2) {
+            // If outputNumberOfChannels>2 but the current route only has two
+            // channels, it means it's actually doing SPDIF under the hood
+            // (usually with a lightning->HDMI adapter).
+            ao->channels = (struct mp_chmap)MP_CHMAP6(FL, FC, FR, SL, SR, LFE);
```
Comment (wm4) [inline audio/out/ao_audiounit.m]: Not sure what that means. With SPDIF I associate compressed passthrough, but we're in the PCM branch here?

File: audio/out/ao_audiounit.m [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+}
+
+static void uninit(struct ao *ao)
+{
+    struct priv *p = ao->priv;
+    AudioOutputUnitStop(p->audio_unit);
+    AudioUnitUninitialize(p->audio_unit);
+    AudioComponentInstanceDispose(p->audio_unit);
+
+    [AVAudioSession.sharedInstance setActive:NO withOptions:AVAudioSessionSetActiveOptionNotifyOthersOnDeactivation error:nil];
```
Comment (wm4) [inline audio/out/ao_audiounit.m]: (Some line breaking would be nice.)

File: audio/out/ao_coreaudio_utils.h [hunk 4]
```diff
@@ -48,8 +48,10 @@ bool check_ca_st(struct ao *ao, int level, OSStatus code, const char *message);
         if (err != noErr) goto label; \
     } while (0)
 
+#if HAVE_COREAUDIO
 void ca_get_device_list(struct ao *ao, struct ao_device_list *list);
```
Comment (wm4) [inline audio/out/ao_coreaudio_utils.h]: No need to ifdef symbols that can be declared unconditionally, even if they're not always defined.

Also, no device selection on iOS?

**Comment by wm4** [comment]
I was kind of hoping iOS could reuse the OSX AudioUnit code. How much of it doesn't actually work on iOS? Since the utils code is full of ifdefs anyway, maybe it would be better to convert existing code to ObjC and ifdef the parts appropriately?

(Not that I like ifdefs much...)

Also, is ao_audiounit.m in any way derived from ao_coreaudio.c? The code looks relatively similar at some points. Can it really be LGPL?

**Comment by wm4** [comment]
Also CC @pigoz for additional review.

**Comment by tmm1** [comment]
> Also, is ao_audiounit.m in any way derived from ao_coreaudio.c? The code looks relatively similar at some points. Can it really be LGPL?

Good catch. I didn't have a header in the file and I copied one from the wrong place before creating the PR.



> I was kind of hoping iOS could reuse the OSX AudioUnit code. How much of it doesn't actually work on iOS? Since the utils code is full of ifdefs anyway, maybe it would be better to convert existing code to ObjC and ifdef the parts appropriately?

Happy to combine them if you don't mind a bunch of `ifdef`s. There is a lot of overlap.

**Comment by pigoz** [comment]
> Happy to combine them if you don't mind a bunch of ifdefs. There is a lot of overlap.

Maybe we can call into ifdeffed functions to leave the logic of the AO as clear as possible, since it looks like the non overlapping parts are of "secondary" importance (buffer/delay querying, channel map negotiation, device negotiation). If we go that route I'd probably isolate the objc parts in their own file since it would make a migration to swift much easier in the future.

> Also CC @pigoz for additional review.

Right now my only gripe is the hardcoded channel negotiation, and the usual 80 cols cosmetic stuff.

**Comment by tmm1** [comment]
Fixed the license, the 80char issues, and the hardcoded channel mapping.

I will leave refactoring overlapping coreaudio/audiounit code to someone else.

File: audio/out/ao_audiounit.m [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+        case kAudioChannelLabel_LeftWide:
+            return MP_SPEAKER_ID_WL;
+        case kAudioChannelLabel_RightWide:
+            return MP_SPEAKER_ID_WR;
+        case kAudioChannelLabel_LFE2:
+            return MP_SPEAKER_ID_LFE2;
+        default:
+            MP_WARN(ao, "Unknown channel mapping for %d\n", ch.channelLabel);
+            return MP_SPEAKER_ID_NA;
+    }
```
Comment (wm4) [inline audio/out/ao_audiounit.m]: Isn't this enum redundant with the list in ao_coreaudio_chmap.c?

**Comment by tmm1** [comment]
> Isn't this enum redundant with the list in ao_coreaudio_chmap.c?

Missed that. Fixed to re-use the existing function.

**Comment by pigoz** [comment]
LGTM 👍 

What do you guys think of renaming this AO to `coreaudio_ios` (this way it'll be consistent with `coreaudio_exclusive`)? Of course this can be done in a future commit before relase

**Comment by tmm1** [comment]
> What do you guys think of renaming this AO to coreaudio_ios (this way it'll be consistent with coreaudio_exclusive)?

Fine by me..

**Comment by pigoz** [comment]
@wm4 can we merge this?

**Comment by wm4** [comment]
Merged.

**Pull request closed without merging**